src/middlewared/middlewared/plugins/apps/crud.py

```diff
@@ -0,0 +1,194 @@
+
+            job.set_progress(60, 'App installation in progress, pulling images')
+            compose_action(app_name, version, 'up', force_recreate=True, remove_orphans=True)
+        except Exception as e:
+            job.set_progress(80, f'Failure occurred while installing {data["app_name"]!r}, cleaning up')
+            for method, args, kwargs in (
+                (compose_action, (app_name, version, 'down'), {'remove_orphans': True}),
+                (shutil.rmtree, (get_installed_app_path(app_name),), {}),
+            ):
+                with contextlib.suppress(Exception):
+                    method(*args, **kwargs)
+
+            raise e from None
+        else:
+            self.middleware.call_sync('app.metadata.generate').wait_sync()
+            job.set_progress(100, f'{data["app_name"]!r} installed successfully')
+            return self.get_instance__sync(app_name)
+
+    @accepts(
+        Str('app_name'),
+        Dict(
+            'app_update',
+            Dict('values', additional_attrs=True, private=True),
+        )
+    )
+    @job(lock=lambda args: f'app_update_{args[0]}')
+    def do_update(self, job, app_name, data):
+        """
+        Update `app_name` app with new configuration.
+        """
+        app = self.get_instance__sync(app_name)
+        config = get_current_app_config(app_name, app['version'])
+        config.update(data['values'])
+        # We use update=False because we want defaults to be populated again if they are not present in the payload
+        # Why this is not dangerous is because the defaults will be added only if they are not present/configured for
+        # the app in question
+        app_version_details = self.middleware.call_sync(
+            'catalog.app_version_details', get_installed_app_version_path(app_name, app['version'])
+        )
+
+        new_values, context = self.middleware.call_sync(
+            'app.schema.normalise_and_validate_values', app_version_details, config, False,
+            get_installed_app_path(app_name),
+        )
+
+        job.set_progress(25, 'Initial Validation completed')
+
+        new_values = add_context_to_values(app_name, new_values, update=True)
+        update_app_config(app_name, app['version'], new_values)
+        job.set_progress(60, 'Configuration updated, updating docker resources')
+        compose_action(app_name, app['version'], 'up', force_recreate=True, remove_orphans=True)
+
+        job.set_progress(100, f'Update completed for {app_name!r}')
+        return self.get_instance__sync(app_name)
+
+    @accepts(Str('app_name'))
+    @job(lock=lambda args: f'app_delete_{args[0]}')
+    def do_delete(self, job, app_name):
+        """
+        Delete `app_name` app.
+        """
+        app_config = self.get_instance__sync(app_name)
+        job.set_progress(20, f'Deleting {app_name!r} app')
+        compose_action(app_name, app_config['version'], 'down', remove_orphans=True)
```

Comment:
Should we be removing images as well which had been pulled and also the named/anonymous volumes which were added?

Reply:
I'll expose that flag, we do add `-v` when we perform `down action` so that would cover the volumes section

Reply:
I see we are using the same flag when stopping the app, that would destroy the volumes as well so we should fix that too